src/lib/utilities/get-workflow-task-failed-event.ts

```diff
@@ -11,18 +12,42 @@ const isFailedTaskEvent = (
   return event.eventType === 'WorkflowTaskFailed';
 };
 
+const isCompletedTaskEvent = (
+  event: WorkflowEvent,
+): event is WorkflowTaskCompletedEvent => {
+  return event.eventType === 'WorkflowTaskCompleted';
+};
+
+const getFailedWorkflowTask = (
+  fullEventHistory: WorkflowEvents,
+): WorkflowTaskFailedEvent | undefined => {
+  const failedWorkflowTaskIndex = fullEventHistory.findIndex(isFailedTaskEvent);
+  const completedWorkflowTaskIndex =
+    fullEventHistory.findIndex(isCompletedTaskEvent);
+
+  if (failedWorkflowTaskIndex < 0) return;
+
+  const failedWorkflowTask = fullEventHistory.find((event) =>
+    isFailedTaskEvent(event),
+  ) as WorkflowTaskFailedEvent;
+
+  if (completedWorkflowTaskIndex < 0 && failedWorkflowTaskIndex >= 0)
+    return failedWorkflowTask;
+
+  // History is sorted in descending order, so index of failed task should be less than index of completed task
+  if (failedWorkflowTaskIndex < completedWorkflowTaskIndex) {
+    return failedWorkflowTask;
+  }
+};
```

Comment:
Not a biggie but I feel like this function could be optimized slightly. Could return early if there is no failed wf task, and remove some iteration. Also this may need some tweaking, I wrote it in GitHub without running/testing.

```suggestion
const getFailedWorkflowTask = (
  fullEventHistory: WorkflowEvents,
): WorkflowTaskFailedEvent | undefined => {
  
  let failedWorkflowTaskIndex = 0;
  let failedWorkflowTask: WorkflowTaskFailedEvent | undefined = undefined;
  let completedWorkflowTaskIndex = 0;

  for (let i = 0; i < fullEventHistory.length; i++) {
    const event = fullEventHistory[i];
    if (isFailedTaskEvent(event) && !failedWorkflowTask) {
      failedWorkflowTaskIndex = i;
      failedWorkflowTask = event;
    }
    
    if (isCompletedTaskEvent(event)) {
      completedWorkflowTaskIndex = i;
    }
  }

  if (!failedWorkflowTaskIndex) return;

  if (completedWorkflowTaskIndex < 0 && failedWorkflowTaskIndex >= 0)
    return failedWorkflowTask;

  // History is sorted in descending order, so index of failed task should be less than index of completed task
  if (failedWorkflowTaskIndex < completedWorkflowTaskIndex) {
    return failedWorkflowTask;
  }
};
```

Reply:
I like findIndex more cuz it's about finding the first occurance (really the latest but since it's reversed history the first) and so you don't need to iterate over the whole event history. But Ill move  if (failedWorkflowTaskIndex < 0) return; above the completedWorkflowTaskIndex so it returns before doing that check if no failed workflow tasks

Reply:
Sounds good!